### PR TITLE
[sca] Batch-capture support for KMAC fixed-vs-random key.

### PR DESCRIPTION
**Pull request by vrozic**
Batch-capture support for KMAC data acquisition. 
Added support for fixed-vs-random key general TVLA.

Signed-off-by: vrozic <vrozic@lowrisc.org>

### CODE REVIEW

File: sw/device/sca/sha3_serial.c [hunk 1]
```diff
@@ -53,8 +53,17 @@ enum {
...
+   * Max number of traces per batch.
+   */
+  kNumBatchOpsMax = 128,
 };
 
+/**
+ * Random message.
+ */
+// static uint8_t random_message[kMessageLength];
+
```
Comment (vogelpi) [inline sw/device/sca/sha3_serial.c:66]: Please remove this.

File: sw/device/sca/sha3_serial.c [hunk 2]
```diff
@@ -471,6 +500,52 @@ static void sha3_serial_single_absorb(const uint8_t *msg, size_t msg_len) {
...
+  uint32_t num_encryptions = 0;
+  uint32_t out[kDigestLength];
+  uint32_t batch_digest[kDigestLength];
+  SS_CHECK(data_len == sizeof(num_encryptions));
+  num_encryptions = read_32(data);
+
+  // This bit may not be needed. Test afterwards
+  for (uint32_t j = 0; j < kDigestLength; ++j) {
+    batch_digest[j] = 0;
+  }
```
Comment (vogelpi) [inline sw/device/sca/sha3_serial.c:518]: Is it needed or not?

File: sw/device/sca/sha3_serial.c [hunk 3]
```diff
@@ -471,6 +500,52 @@ static void sha3_serial_single_absorb(const uint8_t *msg, size_t msg_len) {
...
+    memcpy(kmac_key.share0, batch_keys[i], kKeyLength);
+
+    sca_set_trigger_high();
+    sha3_serial_absorb(batch_plaintexts[i], kMessageLength);
+    sca_set_trigger_low();
+
+    kmac_msg_done();
+    SS_CHECK_DIF_OK(kmac_get_digest(out, kDigestLength));
+
+    for (uint32_t j = 0; j < kDigestLength; ++j) {
```
Comment (vogelpi) [inline sw/device/sca/sha3_serial.c:541]: Do you want to add a comment saying that we basically chain all messages here? In terms of performance, @abdullahvarici figured out that for AES this chaining does have a very big impact on performance. Did you see something similar for KMAC as well?

**Review by vogelpi — COMMENTED** [review]
Thanks for the PR @vrozic ! This looks mostly good. I have some minor questions/comments.

**Comment by vogelpi** [comment]
Hey @vrozic ,
what is the status on this PR? Have you implemented the feedback already and maybe just forgot to force push?

File: sw/device/sca/sha3_serial.c (hunk 3 above)
Comment (abdullahvarici) [reply sw/device/sca/sha3_serial.c:541]: I agree. If we only read and compare a word of the last ciphertext of every batch, the performance would be better.

File: sw/device/sca/sha3_serial.c (hunk 2 above)
Comment (vrozic) [reply sw/device/sca/sha3_serial.c]: I've tested it and it is indeed needed. Comment removed.

File: sw/device/sca/sha3_serial.c (hunk 1 above)
Comment (vogelpi) [reply sw/device/sca/sha3_serial.c]: Okay thanks.

File: sw/device/sca/sha3_serial.c (hunk 3 above)
Comment (vrozic) [reply sw/device/sca/sha3_serial.c:546]: The performance penalty doesn't seem to be very high for KMAC, possibly due to the fewer traces per batch. The capture speed is around 620 traces/s.
I've added the comment in the code explaining that we verify each batch by XORing all outputs in the batch and comparing the result. I would like to keep it as it is for now.

Comment (vogelpi) [reply sw/device/sca/sha3_serial.c:546]: This sounds good to me, thanks for summarizing the conclusion here and for making the comment in the code.

**Review by vogelpi — APPROVED** [review]
LGTM, thanks for initiating this work and then driving it to completion. This is a very nice contribution @vrozic !

**Pull request merged**